v2/bundle/jwtbundle/bundle.go

```diff
@@ -1,73 +1,164 @@
 package jwtbundle
 
 import (
+	"bytes"
 	"crypto"
+	"encoding/json"
+	"errors"
+	"fmt"
 	"io"
+	"io/ioutil"
+	"sync"
 
 	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"gopkg.in/square/go-jose.v2"
 )
 
 // Bundle is a collection of trusted JWT public keys for a trust domain.
 type Bundle struct {
+	mtx            sync.RWMutex
+	trustDomain    spiffeid.TrustDomain
+	jwtSigningKeys map[string]crypto.PublicKey
 }
 
 // New creates a new bundle.
 func New(trustDomain spiffeid.TrustDomain) *Bundle {
-	panic("not implemented")
+	return &Bundle{
+		mtx:            sync.RWMutex{},
+		trustDomain:    trustDomain,
+		jwtSigningKeys: make(map[string]crypto.PublicKey),
+	}
 }
 
 // Load loads a Bundle from a file on disk.
 func Load(trustDomain spiffeid.TrustDomain, path string) (*Bundle, error) {
-	panic("not implemented")
+	bundleBytes, err := ioutil.ReadFile(path)
+
+	if err != nil {
+		return nil, fmt.Errorf("unble to read JWT bundle: %w", err)
+	}
+
+	return Parse(trustDomain, bundleBytes)
 }
 
 // Read decodes a bundle from a reader.
 func Read(trustDomain spiffeid.TrustDomain, r io.Reader) (*Bundle, error) {
-	panic("not implemented")
+	var b bytes.Buffer
+	_, err := b.ReadFrom(r)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read: %v", err)
+	}
+
+	return Parse(trustDomain, b.Bytes())
 }
 
 // Parse parses a bundle from bytes.
-func Parse(trustDomain spiffeid.TrustDomain, b []byte) (*Bundle, error) {
-	panic("not implemented")
+func Parse(trustDomain spiffeid.TrustDomain, bundleBytes []byte) (*Bundle, error) {
+	jwks := new(jose.JSONWebKeySet)
+	if err := json.Unmarshal(bundleBytes, jwks); err != nil {
+		return nil, fmt.Errorf("unable to parse JWK Set: %v", err)
+	}
+
+	bundle := New(trustDomain)
+	for i, key := range jwks.Keys {
+		if err := bundle.AddJWTKey(key.KeyID, key.Key); err != nil {
+			return nil, fmt.Errorf("error adding entry %d of JWK Set: %v", i, err)
+		}
+	}
+
+	return bundle, nil
 }
 
 // TrustDomain returns the trust domain the bundle belongs to.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	b.mtx.RLock()
```

Comment:
Trust domain should be immutable on the bundle so the read lock seems unnecessary.